File: Semestr 3/Paradygmaty Programowania - ćwiczenia/lista4/Scala/lista4_cw.scala
import scala.annotation.tailrec

//zadanie 3

enum BinaryTree[+A] {
  case Node(value: A, left: BinaryTree[A], right: BinaryTree[A])
  case Empty
}

def breadthBT[A](tree: BinaryTree[A]): List[A] = {
  @tailrec
  def bfs(queue: List[BinaryTree[A]], acc: List[A]): List[A] = queue match {
    case Nil => acc.reverse
    case BinaryTree.Empty :: rest => bfs(rest, acc)
    case BinaryTree.Node(value, left, right) :: rest =>
      bfs(rest ++ List(left, right), value :: acc)
  }

  bfs(List(tree), Nil)
}

val tt: BinaryTree[Int] =
  BinaryTree.Node(1,
    BinaryTree.Node(2,
      BinaryTree.Node(4,
        BinaryTree.Empty,
        BinaryTree.Empty
      ),
      BinaryTree.Empty
    ),
    BinaryTree.Node(3,
      BinaryTree.Node(5,
        BinaryTree.Empty,
        BinaryTree.Node(6,
          BinaryTree.Empty,
          BinaryTree.Empty
        )
      ),
      BinaryTree.Empty
    )
  )


// Zadanie 5
class Graph[A](val adjacencyList: A => List[A])

def depthSearch[A](graph: Graph[A], start: A): List[A] = {
  @tailrec
  def dfs(visited: List[A], stack: List[A]): List[A] = stack match {
    case Nil => visited.reverse
    case node :: rest =>
      if (visited.contains(node))
        dfs(visited, rest)
      else {
        val neighbors = graph.adjacencyList(node)
        dfs(node :: visited, neighbors ++ rest)
      }
  }

  dfs(Nil, List(start))
}

val g = new Graph[Int](node =>
  node match {
    case 0 => List(4, 3, 1)
    case 1 => List(0)
    case 2 => List(3)
    case 3 => List(0, 2)
    case 4 => List(0)
    case _ => Nil
  }
)

@main
def main(): Unit = {
  println("\nZadanie 3:")
  val zadanie3 = breadthBT(tt)
  println(s"[${zadanie3.mkString("; ")}]")

  println("\nZadanie 5:")
  val zadanie5 = depthSearch(g, 4)
  println(s"[${zadanie5.mkString("; ")}]")
}

File: Semestr 3/Paradygmaty Programowania - ćwiczenia/lista04/OCaml/lista4_cw.ml
(*zadanie 2*)

let rec f x = f x

type 'a bt = Empty | Node of 'a * 'a bt * 'a bt

(*zadanie 3*)

let breadthBT tree =
  let rec bfs queue acc = match queue with
    | [] -> List.rev acc
    | Empty :: rest -> bfs rest acc
    | Node (value, left, right) :: rest ->
      bfs (rest @ [left; right]) (value :: acc)
  in
  bfs [tree] []
;;

let tt =  Node(1,
            Node(2,
              Node(4,
                Empty,
                Empty
                ),
              Empty
              ),
            Node(3,
              Node(5,
                Empty,
                Node(6,
                  Empty,
                  Empty
                  )
                ),
              Empty
              )
          )
;;

(*zadanie 5*)

type 'a graph = Graph of ('a -> 'a list)

let rec depthSearch graph start =
  let rec dfs visited stack = match stack with
    | [] -> List.rev visited
    | node :: rest ->
      if List.mem node visited then
        dfs visited rest
      else
        let neighbors = match graph with
          | Graph(g) -> g node
        in
        dfs (node :: visited) (neighbors @ rest)
  in
  dfs [] [start]
;;

let g = Graph (fun node ->
  match node with
  | 0 -> [4; 3; 1]
  | 1 -> [0]
  | 2 -> [3]
  | 3 -> [0; 2]
  | 4 -> [0]
  | _ -> []
)
;;

let () =
  Printf.printf "\nZadanie 3: \n";
  let zadanie3 = breadthBT tt in 
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int zadanie3));

  Printf.printf "\nZadanie 5: \n";
  let zadanie5 = depthSearch g 4 in 
  Printf.printf "[%s]\n" (String.concat "; " (List.map string_of_int zadanie5));
;;